crates/driver/src/domain/competition/settled/mod.rs

```diff
@@ -0,0 +1,238 @@
+use {
+    super::{
+        auction,
+        order::{self, Side},
+    },
+    crate::{domain::eth, util::conv::u256::U256Ext},
+    number::conversions::big_rational_to_u256,
+};
+
+/// Settlement built from onchain calldata.
+#[derive(Debug, Clone)]
+pub struct Settlement {
+    trades: Vec<Trade>,
+}
```

Comment:
How important is the fact that this is built from onchain data? Given that we are defining and require all internal structs publicly it doesn't really matter, does it? We could as well instantiate the `Vec<Trade>` from a `competition::Solution`, right?

In fact, I think the `settled::Solution` namespace is somewhat misleading because we are planning to use this datatype way before the solution has been settled. If it were important, I'd make this be constructable from a `eth::Tx` type (which would make it very clear that you need an ethereum transaction to create this type) and leave the `Trade` and `Prices` struct private. But then I believe we would have to re-implement the whole decoding logic (which would make this PR explode).

Could we move this struct into `competition/solution/score.rs` and call `new` from the `Solution` struct (maybe in the constructor after the fee has been applied or just in time in `Solution::Score()`? Unless I'm missing something about the importance of using the encoded settlement I think this would be most straightforward (and inline with #1494)

Reply:
> We could as well instantiate the Vec<Trade> from a competition::Solution, right?

I expect that to happen with `SettlementEncoder` rewrite.

> But then I believe we would have to re-implement the whole decoding logic (which would make this PR explode).

This is done in [autopilot](https://github.com/cowprotocol/services/pull/2436). However, here I took a small shortcut to keep the PR smaller.

> Unless I'm missing something about the importance of using the encoded settlement

Definitely doable however requires knowing in details how SettlementEncoder works (like this one https://github.com/cowprotocol/services/blob/main/crates/solver/src/settlement/settlement_encoder.rs#L267-L275). To move faster, I reused already prepared `EncodedSettlement` to not deal with this at this time (I can work on this in following weeks).

To me it's not important source from which the settled::Settlement is built, It's just important to have the same semantics as `EncodedSettlement` (and now this is guaranteed by SettlementEncoder which is battle tested 😄 ). This allows the driver/autopilot/solver-rewards to use 100% exact math for surplus/fees calculations, which gives me high confidence that the code is correct and save time on bug fixes (which happened a lot with protocol fee calculation and misalignment with solver rewards).

Reply:
> To me it's not important source from which the settled::Settlement is built, It's just important to have the same semantics as EncodedSettlement (and now this is guaranteed by SettlementEncoder which is battle tested 😄 ).

Well, `competition::Solution` is also battle-tested since it's what feeds into the encoded Settlement (if it has bugs so will the encoded settlement). I agree that using the same math is important but since you are using newly introduced types you would have the same guarantees if you create the struct from `competition::Solution` with the additional advantages of

1. Less confusion about why we use a `settled::Settlement` before it's even settled
2. No dependency on a struct that's supposed to go away (this PR will make it harder to remove the encoded settlement)

What are the downsides of my suggestion? I don't think it's that much work to implement as the meat of this PR (all the math inside `settled::Settlement` can stay, it just changes how to instantiate this class.

Reply:
Done.